theano/compile/function_module.py

```diff
@@ -483,6 +488,60 @@ def __copy__(self):
                 there.data = here.data
         return cpy
 
+    def copy(self, share_memory=False):
+        """
+        Copy this function. Copied function will have separated maker and fgraph
+        with original function. User can choose whether to separate storage by 
+        changing the share_memory arguments
+        ---------------------
+        Params:
+            share_memory -- { boolean } Default is False. When True, two function
+            share intermediate storages(storages except input and output storages)
+        ---------------------
+        Returns:
+            func -- Copied theano.Function
+        """
+
+        if not share_memory:
+            return self.__copy__()
+        else:
+            # copy SymbolocKits
+            ins, outs = copy.deepcopy([self.maker.inputs, self.maker.outputs])
+
+            # copy fgraph and get memo
+            maker = self.maker
+            fg_cpy, memo = maker.fgraph.clone_get_equiv(attach_feature=False)
+
+            # use copied ins, outs and fgraph to init a maker
+            new_maker = maker.__class__(inputs=ins, outputs=outs, mode=maker.mode,
+                                        fgraph=fg_cpy, profile=maker.profile,
+                                        accept_inplace=maker.accept_inplace,
+                                        function_builder=maker.function_builder,
+                                        on_unused_input=maker.on_unused_input)
+
+            # construct new storage_map that map new variable to old storage
+            # so that the ensuing function shares storage with the original one
+            new_storage_map = {}
+            storage_map = self.fn.storage_map
+            for key in storage_map.keys():
+                # output_storages should not be shared
+                if key not in self.maker.fgraph.outputs and memo.has_key(key):
```

Comment:
Why memo would not have key in it? When this happen? I think it should never happen. If it is the case, remove the corresponding condition. Otherwise, put a comment that tell why this happen.

I don't understand why we should not share the output storage. If Theano gc is enabled, it will be emptied so we don't care (and this share_memory would be useless anyway). If Theano gc is disabled, sometimes we would like to reuse it. But currently, there is another mechanism that will prevent that, even without this.

But I think for now, it is safer to skip it like you do. But can add a comment like this, so that we have the information for later:

TODO: We could share the output storage, but we must make sure 2 different function call won't override each other values. This is already done elsewhere, so to reuse it the user would need to use Out(var, borrow=True) and maybe the mutable=True flag too. But to be safe for now as it isn't documented and we aren't sure it is well tested, we don't share the part of the storage_map.